### ISSUE TITLE
Semantic versioning convention

### ISSUE THREAD

**Issue by czlee** [czlee/tabbycat#216]
Opening a discussion to formalize semantic versioning, with an initial goal of determining whether it's appropriate to release Abyssinian as 1.0.0 rather than 0.7.0. Relevant web pages:
- http://semver.org/
- http://programmers.stackexchange.com/questions/255190/how-does-semantic-versioning-apply-to-programs-without-api
- https://github.com/mojombo/semver/issues/90

Since semantic versioning requires a "public API" and Tabbycat is for the end-user, _i.e._ no other software is dependent on Tabbycat, if we want to use semantic versioning properly, we need some idea of what it means for us. The second and third links above are discussions towards that end in other contexts.

Here's a proposed definition/substitute for "public API", with criteria for backwards incompatibility:
- **Database**
  - if it cannot be migrated forwards or backwards using the standard migration function without user-input defaults
  - if migration forwards would entail losing data or require reformatting data
- **Management commands**
  - if a command that used to work no longer works
- **GUI**
  - if there is a major change to the workflow of any user
- **Tournament data importer, including tournament configuration**
  - if files that used to work would no longer work. 
  - However, with tournament configuration, I'd imagine that Tabbycat could in most cases detect deprecated settings and interpret them in any new framework with a warning message.

Major refactoring I'd like done before 1.0.0, with counterarguments:
- Move all data utilities commands into management commands
  - Could argue that data/utils commands aren't technically part of the "public API", so this is an addition not a change.
- Overhaul tournament configuration structure.
  - This is mainly internal and needn't necessarily change the way configuration is stored in the database. But it'd be nice to have that flexibility not to worry about backwards compatibility yet.
- Restructure the database to a structure more amenable for performance
  - This might be too tall an ask and better saved for 2.0.0!

Major refactoring that would be nice to do, but I suppose don't affect the "public API":
- Restructure into multiple apps.
- Extract all standings sorters into their own library. This would require replacing the "australs"/"nz" configuration with something more customizable, but that could be detected as discussed above.

When this issue is closed, the definition of "public API" for semantic versioning purposes should be moved to the wiki.

**Comment by philipbelesky** [comment]
Don't have any strong thoughts of this, but my general preference being:

- X.Y.Y releases encompassing a cohesive set of cumulative functionality that has been rigorously. Functionality can be defined as a significant new or modified user-facing workflow, or model changes that require a difficult or impossible schema migration. 
- Y.X.Y releases for changes where a minor database migration is necessary, or a workflow (as defined above) is changed/modified.
- Y.Y.X releases for bug fixes, cosmetic changes, optimisation, and/or minor features that do not violate any of the above.

So, as applies:

- I think the changes over the last few months are enough to cumulatively bump up to a 1.0
- Refactoring the database structure would definitely be a 2.0, due to the painful migration path. Redoing the configuration setup could also potentially merit this, as might something along the lines of BP support, or a shadow draw system.
- Any small schema changes or features (say for WADL) would be a 1.1 or similar
- Most likely candidates for a 1.0.1 would be hotfixes

Most of our development happens at limited points in the year and/or is beta tested at actual tournaments. As such, features and fixes usually come in bundles. So I would expect several minor version bumps in the first half of the year (minor features debuting at smaller tournaments) and perhaps a major version bump after a big tournament where extensive testing has been done and at least one big feature or schema change has occurred.

So thats kind of a pretty rough set of semantics. I'm more concerned about ensuring releases are stable though, and ensuring the release increments work with how the pace of our development/testing plays out in practice.

**Comment by czlee** [comment]
I'm a little confused about the correlation with testing or the size of the tournament. Presumably all our releases should be tested before release. I mean, we don't really have a robust testing cycle beyond beta-testing at tournaments (skipping both alpha and RC), but anything that we don't have confidence in shouldn't be released, whether it's a patch or major.

Because of the infrequency of tournaments the nature of our testing cycle is (by software standards) weird... Maybe we should have some sort of intermediate (beta) release as a matter of practice, in particular if there are people other than us potentially willing to shoulder some risk in return for more features (or being kind).

Otherwise, that first set of bullet points sound roughly equivalent to my proposal...

We should bump to 1.0.0 if, and only if, we don't foresee doing any backwards-incompatible changes doing any time soon. I think you can argue that a lot of the refactoring I'd still like to do isn't "backwards-incompatible" per se, but if we think getting the database structure right will happen before too long, then we should continue with 0.x.x releases. Also, I realized that that "restructure into multiple apps" will (contrary to my first message) be a painful migration, because Django inserts app names into table names (`debate_debateadjudicator` not `debateadjudicator`), so the same applies to that plan. It doesn't really make much sense to move to 1.0.0 if we think we'll hit 2.0.0, say, within the next year, since releasing 1.0.0 is (according to semver.org) a statement that we think the public API is now well-defined.

**Comment by philipbelesky** [comment]
At the point our API is defined as both the schema and the functionality, I think it's well defined in that it has been used (in more or less the current form) at a number of tournaments, so its a bit of a value judgment. I wouldn't mind a release cycle that has a major number bump each year, but thats probably a subjective value judgment that I don't feel strongly about. 

With regards to the correlation with testing, I think to a certain extent what counts as a significant feature/functionality should be realistically calibrated to the release cycle that is defined by tournament schedules. IE at the point the software has some new features and has been evaluated thoroughly, that should constitute a minor version bump. Again, subjectively, I would like there to be at least a couple of minor version bumps over the course of the year, rather than the current model where things sit in /develop/ for half the year before being folded in all at once. 

So, not sure I disagree on anything (aside from 1.0 readiness) but want to ensure whatever definition is adopted ensures a slightly steadier release cycle, particularly if development slows.

**Comment by czlee** [comment]
Oh, by "well-defined" I meant we're confident it should look like what it currently is. If we think its current form is what it should be, such that we should be very careful about changing it, then we should be on 1.0.0. If we are still of the mindset that we can, and want to be able to, change things without regard for backwards compatibility, then we should stay on 0.x.x. This is the only value judgement that I think is relevant.

According to the semantic versioning specs, the 1.0.0-to-2.0.0 decision is different from 0.0.0-to-1.0.0, sorry that part was confusing in my last comment: what I meant was a 0-to-1 bump makes no sense if you *know* what the 1-to-2 bump is going to be, because if you have a 1-to-2 bump in mind, your public API is (by definition) not well-defined.

I plan to do the configuration overhaul as my next task, but I think that can be made backwards compatible. Things that cannot be backwards compatible are (1) database restructure and (2) apps restructure. If we know we want the freedom to do this, we should stay on 0.x.x, but if it's too far in the distance (like basically it's in our too-hard basket, so it'll likely be at least a year away and we'll be doing other things in the meantime) then I'm fine with going to 1.0.0 now.

Can we enforce that steadier release cycle by making greater use of beta releases? I think our version bumps should be related only to the functionality added/fixed, as per semantic versioning, and any relationship to tournament size or time should be coincidental only. But I agree we need to get into the habit of releasing more often, and I was thinking releasing a version for advanced users willing to try things out for us might be a way of achieving this. We could turn the Facebook group into a beta testing programme! (I'm only half-joking.)

**Comment by philipbelesky** [comment]
Ok, with regard to 1.0.0 vs 2.0.0, it's probably best to wait then. I have some time blocked out (probably towards the end of the year) to finish off the WADL spec, and those changes could be backwards-incompatible. At the same time I'll probably do some database/app restructures to try and get rid of the performance blocks. And with that in mind, feel free to make the config overhaul backwards incompatible also. 

So maybe plan to revisit things towards the end of the year? And release current ```master``` as 0.7.0?

**Comment by czlee** [comment]
Sounds good!

Do the criteria for backwards compatibility in my first comment sound sensible to you?

**Comment by philipbelesky** [comment]
Yes, all good

**Comment by czlee** [comment]
I realized that all of the data migrations I've written so far aren't reversible, so break our criteria for backwards compatibility. In some cases making them reversible is trivial (we just make the reverse operation a noop), so we could do that. But it seems more sensible to me just not to support reverse migrations—in general we should expect users to upgrade and never downgrade, and if they do want to downgrade then they're on their own.

I propose modifying our backwards compatibility definition as follows: (~~delete~~, **add**)

* Database
    * if it cannot be migrated forwards ~~or backwards~~ using the standard migration function ~~without user-input defaults~~
    * if migration forwards would entail losing data **without replacement** ~~or require reformatting data~~

By "without replacement" I mean that data is lost that can't be inferred from new fields or something like that. This is actually quite a high bar that would almost never be met, but maybe that's how it should be.

**Comment by philipbelesky** [comment]
Makes sense. I think the 'forward without replacement' think occasionally happens — such as with venue preferences, in which case that makes sense as a breaking version.

Maybe there's a case for making migrations backwards compatible within a minor version, or within a major version, but I don't think this is ever likely to be useful in practice except for reverting a bad hotfix or something.

**Comment by czlee** [comment]
Oh right, in that we didn't write a data migration? Yeah, that's true I guess.

Another option is to ditch our adherence to semantic versioning entirely. I'm not totally opposed to this, given that we don't have an API (yet), and it's not really clear to me that any of the criteria we currently have failing would necessarily be so disruptive that a major version bump would be an important signal to users. Basically, except for migrations, all of the criteria concern input from humans, which isn't as disruptive to break as interaction with other software.

**Comment by philipbelesky** [comment]
So the problem is that if we only care about forwards migration compatibility that we will never get to major version bumps? Or that our 'API' definition is too loose  now?

It seems like the versioning convention was already a pretty loose fit. I'd imagined that a major version bump would only come around from the *if there is a major change to the workflow of any user* or if there was some serious schema change that required starting from scratch. In which case our major/minor version bumps (in practice) align with what the general expectation of software updates are.

**Comment by czlee** [comment]
We can always bump versions (major or minor) whenever we want, semantic versioning only prohibits us from _not_ bumping versions in some cases. So the concern (to the extent there is any) is that our convention would have us incrementing major versions more than seems intuitively appropriate, which would happen if the criteria for backwards compatibility were too demanding.

That all said, I just realised that we haven't actually faced this yet. Pre-1.0 you can change whatever you want without major version bumps (by definition pre-1.0 is unstable), and in 1.1 we didn't remove anything. So our past practice doesn't really provide any indicator.

But I still suspect our current criteria (to avoid a major version bump) will turn out to be too demanding. The only hard criterion I can think of is that if forward migrations won't work, it should definitely be a major version bump. Management command compatibility probably isn't worth stressing over; changes should just be signalled in release notes and via useful error messages. The "importer files won't work" criterion is almost certainly too strict—if we, say, add a new required field to a model, then importer files obviously have to be modified accordingly, so long as it's easy enough I think this is fine. Any criterion about "major change to workflows" is fine in principle, but it's not really measurable.

Basically I'm wondering if all this gymnastics about what exactly "constitutes" a public API is a bit silly, and if we should just drop adherence to semantic versioning altogether—at least before we _actually_ implement an API that third-party software (_e.g._ registration systems, feedback systems) can hook into.

**Comment by philipbelesky** [comment]
Sure, I think we can drop the sem ver moniker, or perhaps cast it as 'sem-ver-like' — ie say the only strict criteria is forward migrations not working. I'm pretty ambivalent about either option.
